tough/src/editor/targets.rs

```diff
@@ -0,0 +1,536 @@
+                    delegated_role.keyids.extend(keyids.clone());
+                }
+            }
+            for delegated_role in self.new_roles.get_or_insert(Vec::new()).iter_mut() {
+                if delegated_role.name == role {
+                    delegated_role.keyids.extend(keyids.clone());
+                }
+            }
+        }
+        Ok(self)
+    }
+
+    /// Removes a key from delegations keyids, if a role is specified only removes it from the role
+    pub fn remove_key(&mut self, keyid: &Decoded<Hex>, role: Option<&str>) -> Result<&mut Self> {
+        let delegations = self.delegations.as_mut().context(error::NoDelegations)?;
+        // If a role was provided remove keyid from the delegated role
+        if let Some(role) = role {
+            for delegated_role in &mut delegations.roles {
+                if delegated_role.name == role {
+                    delegated_role.keyids.retain(|key| keyid != key);
+                }
+            }
+        } else {
+            delegations.keys.remove(keyid);
+        }
+        Ok(self)
+    }
+
+    /// Delegates `paths` to `targets` adding a `DelegatedRole` to `new_roles`
+    pub fn delegate_role(
+        &mut self,
+        targets: Signed<DelegatedTargets>,
+        paths: PathSet,
+        key_pairs: HashMap<Decoded<Hex>, Key>,
+        keyids: Vec<Decoded<Hex>>,
+        threshold: NonZeroU64,
+    ) -> Result<&mut Self> {
+        self.add_key(key_pairs, None)?;
+        self.new_roles
+            .get_or_insert(Vec::new())
+            .push(DelegatedRole {
+                name: targets.signed.name,
+                paths,
+                keyids,
+                threshold,
+                terminating: false,
+                targets: Some(Signed {
+                    signed: targets.signed.targets,
+                    signatures: targets.signatures,
+                }),
+            });
+        Ok(self)
+    }
+
+    /// Removes a role from delegations if `recursive` is `false`
+    /// requires the role to be an immediate delegated role
+    /// if `true` removes whichever role eventually delegates 'role'
+    pub fn remove_role(&mut self, role: &str, recursive: bool) -> Result<&mut Self> {
+        let delegations = self.delegations.as_mut().context(error::NoDelegations)?;
+        // Keep all of the roles that are not `role`
+        delegations
+            .roles
+            .retain(|delegated_role| delegated_role.name != role);
+        if recursive {
+            // Keep all roles that do not delegate `role` down the chain of delegations
+            delegations.roles.retain(|delegated_role| {
+                if let Some(targets) = delegated_role.targets.as_ref() {
+                    targets.signed.delegated_role(role).is_err()
+                } else {
+                    true
+                }
+            });
+        }
+        Ok(self)
+    }
+
+    /// Adds a role to `new_roles`
+    pub fn add_role(
```

Comment:
Could we clarify the difference between this method and `delegate_role()` to make it clear which method a user should use?

Reply:
Based on the comment, is the only difference between the two methods that this one fetches the metadata from a URL? If so, let's make the methods allude to that fact.

Also - let's add to the docstring a few words as to why and when I would use either of these